### ISSUE TITLE
More on FastList is not thread-safe

### ISSUE THREAD

**Issue by kooroshvak** [brettwooldridge/HikariCP#955]
I am getting similar Exception to issue #591. I am pretty sure my threads are not sharing the same connection at least not directly. As a matter of fact, the exceptions happen when two threads are trying to close their current connection and get a new connection from the pool. The application has over 512 threads running on a 16 cores machine. The process is IO intensive and CPU usage is not that high. Here are the two exceptions, the exceptions are coming from two different threads:

`Exception in thread "MainWorker-171" java.lang.ArrayIndexOutOfBoundsException: -1
	at com.zaxxer.hikari.util.FastList.remove(FastList.java:136)
	at com.zaxxer.hikari.pool.ProxyConnection.untrackStatement(ProxyConnection.java:169)
	at com.zaxxer.hikari.pool.ProxyStatement.close(ProxyStatement.java:70)
	at com.xxx.fsmon.summary.DirectorySummaryDAO.close(DirectorySummaryDAO.java:194)
	at com.xxx.fsmon.summary.DirectorySummaryDAO.renewConnectionToDb(DirectorySummaryDAO.java:159)
	at com.xxx.fsmon.directory.ThreadByDir.DirectoryRunnableExtractor.extract(DirectoryRunnableExtractor.java:215)
	at com.xxx.fsmon.directory.ThreadByDir.DirectoryRunnableExtractor.run(DirectoryRunnableExtractor.java:454)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)`

and

`[MainWorker-488] ERROR com.xxx.fsmon.summary.FileSummaryDAO: Could not prepare the statements!
java.lang.ArrayIndexOutOfBoundsException: -1
        at com.zaxxer.hikari.util.FastList.add(FastList.java:78)
        at com.zaxxer.hikari.pool.ProxyConnection.trackStatement(ProxyConnection.java:189)
        at com.zaxxer.hikari.pool.ProxyConnection.prepareStatement(ProxyConnection.java:312)
        at com.zaxxer.hikari.pool.HikariProxyConnection.prepareStatement(HikariProxyConnection.java)
        at com.xxx.fsmon.summary.FileSummaryDAO.<init>(FileSummaryDAO.java:116)
        at com.xxx.fsmon.summary.DirectorySummaryDAO.renewConnectionToDb(DirectorySummaryDAO.java:171)
        at com.xxx.fsmon.directory.ThreadByDir.DirectoryRunnableExtractor.extract(DirectoryRunnableExtractor.java:215)
        at com.xxx.fsmon.directory.ThreadByDir.DirectoryRunnableExtractor.run(DirectoryRunnableExtractor.java:454) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)`

Could this be caused by a race condition in HikariCP code where one thread closes a connection, but has not finished closing all its statements and the other thread somehow gets that connection and is trying to open a new prepareStatement on it?

I am going to synchronize my renewConnectionToDb() method and see if the exceptions go away.

Any suggestion is greatly appreciated.

#### Environment
```
HikariCP version: 2.6.3
JDK version     : 1.8.0_11
Database        : MySQL
Driver version  : 6.0.5
```
-----------------------------------------------------------------------------------------

**Comment by brettwooldridge** [comment]
This looks like a legitimate cross-core  L2 cache issue. I'll fix right away.

**Comment by guidomedina** [comment]
@kooroshvak why such an out of date JDK 8?

JDK 8 introduced so many bugs before JDK 8u25, I would try to reproduce the issue using an updated version.

**Comment by kooroshvak** [comment]
Thanks for prompt response.

FYI, since I have synchronized my method, renewConnectionToDb(), I haven't seen this issue yet.

As far as upgrading my JDK, I will look into it. Thanks for the suggestion.

**Comment by brettwooldridge** [comment]
It is my belief that synchronizing the FastList.clear() method will be sufficient to synchronize the L2 caches across physical CPUs. This change will be published in the next few hours.

**Comment by brettwooldridge** [comment]
@kooroshvak This actually looks to be different synchronization issue, not ``FastList``.  Are statements are prepared on one thread, but closed on a different thread?

I *am* curious about the code pattern implied by the stacktrace above.  ``DirectorySummaryDAO.renewConnectionToDb()`` is responsible for creating new instances of ``FileSummaryDAO`` (which runs SQL), as well as closing instances of ``DirectorySummaryDAO``?

**Comment by kooroshvak** [comment]
No, there is a parent-child relationship between DirectorySummaryDAO and FileSummaryDAO objects. They share the same DB connection, but create their our specific SQL statements. When closing DirectorySummaryDAO, it would close all its statements and then calls the corresponding close() method on the FileSummaryDAO which do the same on its own SQL statements. Now that I am looking at my code closely, I believe I have found a bug, since I am closing the DB connection on DirectorySummaryDAO before calling FileSummaryDAO to close its SQL statements. Sorry, it looks like the bug is on my side. My apologies.

**Comment by brettwooldridge** [comment]
That alone should not cause the issue.  The way the [code is written](https://github.com/brettwooldridge/HikariCP/blob/0ab35e744553914f2dda5a34b659b31c49058d41/src/main/java/com/zaxxer/hikari/pool/ProxyStatement.java#L65) ensures that a Statement is only closed once.  So, whether all of the statements are closed as a side-effect of closing the connection or not *should* have no effect.  Later calls to close those same statements should be no-ops.

The issue I did find was that because the *isClosed* flag was neither volatile nor atomic, access to the same state via different cores *could* result in one core seeing the statement as still open.

The higher the core count, the higher the likelihood of that happening.

The change I committed synchronizes on the isClosed state and ensures cache-line flushing.
